core/src/main/java/com/graphhopper/routing/weighting/custom/ExpressionBuilder.java

```diff
@@ -0,0 +1,379 @@
+
+    // Introduce a dynamic cache to remember different Weighting classes, but throw away less frequently used classes.
+    // Use accessOrder==true to remove oldest accessed entry, not oldest inserted.
+    private static final int DYN_CACHE_SIZE = Integer.getInteger("graphhopper.custom_weighting.dynamic_cache_size", 1000);
+    private static final Map<String, Class> DYN_CACHE = Collections.synchronizedMap(
+            new LinkedHashMap<String, Class>(DYN_CACHE_SIZE, 0.75f, true) {
+                protected boolean removeEldestEntry(Map.Entry eldest) {
+                    return size() > DYN_CACHE_SIZE;
+                }
+            });
+
+    /**
+     * This method compiles a new subclass of CustomWeightingHelper composed from the provided CustomModel caches this
+     * and returns an instance.
+     */
+    static CustomWeightingHelper create(CustomModel customModel, EncodedValueLookup lookup,
+                                        double globalMaxSpeed, double maxSpeedFallback, DecimalEncodedValue avgSpeedEnc) {
+        Java.CompilationUnit cu;
+        try {
+            String key = customModel.toString() + ",global:" + globalMaxSpeed + ",fallback:" + maxSpeedFallback;
+            if (key.length() > 400_000) throw new IllegalArgumentException("Custom Model too big: " + key.length());
+
+            Class clazz = CACHE.get(key);
+            if (DYN_CACHE_SIZE > 0 && clazz == null)
+                clazz = DYN_CACHE.get(key);
+            if (clazz == null) {
+                HashSet<String> priorityVariables = new LinkedHashSet<>();
+                List<Java.BlockStatement> priorityStatements = createGetPriorityStatements(priorityVariables, customModel, lookup);
+                HashSet<String> speedVariables = new LinkedHashSet<>();
+                List<Java.BlockStatement> speedStatements = createGetSpeedStatements(speedVariables, customModel, lookup, globalMaxSpeed, maxSpeedFallback);
+                // Create different class name, which is required only for debugging.
+                // TODO does it improve performance too? I.e. it could be that the JIT is confused if different classes
+                //  have the same name and it mixes performance stats. See https://github.com/janino-compiler/janino/issues/137
+                long counter = longVal.incrementAndGet();
+                String classTemplate = createClassTemplate(counter, priorityVariables, speedVariables, lookup, customModel);
+                cu = (Java.CompilationUnit) new Parser(new org.codehaus.janino.Scanner("source", new StringReader(classTemplate))).
+                        parseAbstractCompilationUnit();
+                cu = injectStatements(priorityStatements, speedStatements, cu);
+                SimpleCompiler sc = createCompiler(counter, cu);
+                clazz = sc.getClassLoader().loadClass("com.graphhopper.Test" + counter);
```

Comment:
Can we use a bit more meaningful name than `Test`? Something like `JaninoAutoGeneratedCustomWeighting` or something (if this is what it is)?